### PR TITLE
Exclude projects for aspx/cshtml/vbhtml files

### PR DESCRIPTION
**Pull request by tmeschter**
When mapping an `IVsHierarchy` for a project node in Solution Explorer
to a Roslyn `AbstractProject` we try to match the properties on the
node's hierarchy against properties on each project's hierarchy. In the
case of a Web Application Project we may still end up with multiple
matching Roslyn projects: one for the "main" project, and one for each
open aspx/cshtml/vbhtml file. We're only interested in the main
project, so this change excludes any projects with a
`ContainedDocument`. This may involve checking a lot of documents, so
we only do it if we must.

**Customer scenario**

Customer opens any Web Application project, and then opens one or more .aspx, .cshtml, or .vbhtml files. If they then search in the Solution Explorer window, VS will very likely crash.

**Bugs this fixes:**

Fixes [VSO bug 441171](https://devdiv.visualstudio.com/DevDiv/_workitems?id=441171).

**Workarounds, if any**

None.

**Risk**

Low; we use this approach elsewhere to filter out Roslyn projects associated with an open aspx/cshtml/vbhtml file.

**Performance impact**

There will be little or no perf impact. This code path will generally only be executed when the contents of the Solution Explorer window change; e.g. during solution load or while searching in Solution Explorer. The new check (for projects with a `ContainedDocument`) is an additional cost, so it is only utilized if the existing checks are insufficient to narrow us down to a single project.

**Is this a regression from a previous update?**

Yes.

**Root cause analysis:**

In order to support showing analyzers and diagnostics in the Solution Explorer for both the old and new project systems it was necessary to update the code that maps `IVsHierarchy` nodes to Roslyn projects, and the set of checks used to determine if a node matched a project changed. These updated checks did not account for the Roslyn projects created when .aspx, .cshtml, or .vbhtml files are opened in VS.

Issue https://github.com/dotnet/project-system/issues/1607 covers adding a test to cover this scenario (among others related to the Dependencies node).

**How was the bug found?**

Automated testing.

### CODE REVIEW

**Comment by tmeschter** [comment]
@dotnet/project-system @dotnet/roslyn-ide @srivatsn @Pilchie Please review.

File: src/VisualStudio/Core/SolutionExplorerShim/HierarchyItemToProjectIdMap.cs [hunk 1]
```diff
@@ -43,45 +44,62 @@ public bool TryGetProjectId(IVsHierarchyItem hierarchyItem, string targetFramewo
...
-                    .SingleOrDefault();
+            // If we have multiple candidates then we might be dealing with Web Application Projects. In this case
+            // there will be one main project plus one project for each open aspx/cshtml/vbhtml file, all with
+            // identical properties on their hierarchies. We can find the main project by excluding everything
+            // with a contained document. We may end up looking at lots of documents, so we do this check only if we
+            // must.
+            if (candidateProjects.Length > 1)
+            {
+                candidateProjects = candidateProjects
+                    .Where(p => !p.GetCurrentDocuments().Any(doc => doc is ContainedDocument))
```
Comment (Pilchie) [inline src/VisualStudio/Core/SolutionExplorerShim/HierarchyItemToProjectIdMap.cs]: This forces evaluation of all of the rest of the projects even if you've found one with no `ContainedDocuments` - I'd expand this into a foreach loop that breaks early.

**Comment by tmeschter** [comment]
@jasonmalinowski No, this shouldn't impact multi-targeting class libs or anything like that. Most of the time we'll filter down to one project before we even need to check for `ContainedDocuments` so we won't even need to do that.

**Comment by tmeschter** [comment]
@Pilchie I've made an update to address your concern.

**Comment by tmeschter** [comment]
@dotnet-bot test ubuntu_14_debug_prtest please

**Comment by tmeschter** [comment]
Tagging @MattGertz for Escrow approval (pending tests).

**Comment by natidea** [comment]
Shiproom Approved to merge via [441171](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/441171)

**Pull request merged**